Add CLIP model

Kamino666 opened the pull request:
CLIP is an interesting model with features that have achieved excellent results in several downstream tasks. So I think adding this model will make this repo better.
[paper](https://arxiv.org/abs/2103.00020), [code](https://github.com/openai/CLIP)

In general, I added the wrapping code for feature extraction in `models/clip` , modified the corresponding part in `main.py` and `utils/utils.py`, and made a Colab notebook. Most of the documentation has also been completed, except for 1. the model figures and 2. the contents of the Colab notebook. 

+ For the first unfinished content, I don't have the original file of the figure, therefore you may need to make one :laughing:. It should look almost the same as ResNet.

+ For the second, there is a figure in the notebook that needs to change, and it uses my forked repo currently. Therefore you may need to modify it and save to your Google Drive. 

I have done test in Colab and my CPU&Windows machine.

Note that `show_pred` does not work in this model because it's trained with contrastive task but not classification task. (I have added this note in the documentation)

v-iashin commented on models/clip/extract_clip.py:
```diff
@@ -0,0 +1,175 @@
+                vid_feats.extend(batch_feats.tolist())
+                # show predicitons on imagenet dataset (might be useful for debugging)
+                # if self.show_pred:
+                #     logits = classifier(batch_feats)
```
we need to find a way :)

Kamino666 replied:
I think we can ask the user to provide a sentence and then calculate the similarity of the video to it. Just like the first [CLIP example](https://github.com/openai/CLIP/blob/main/README.md#usage), since this task should be more in line with the training method. This would import an extra configuration item, but it shouldn't be a big problem.
Of course we could also use Zero-Shot Prediction as you mentioned, since apparently they already offer this as a demonstration.

v-iashin commented on models/clip/extract_clip.py:
```diff
@@ -0,0 +1,175 @@
+            model, preprocess = clip.load("RN101", device=device)
+        elif self.feature_type == 'CLIP-RN50':
+            model, preprocess = clip.load("RN50", device=device)
+        elif self.feature_type == 'CLIP-custom':
```
Let's remove the custom model support for now :/

Kamino666 replied:
I will use this feature in my research, but it doesn't matter.

v-iashin replied:
Oh, I see. Actually, that's ok. Let's leave it. I just realized that you put the doc already! Great.

v-iashin commented on utils/utils.py:
```diff
@@ -244,6 +244,9 @@ def build_cfg_path(feature_type: str) -> os.PathLike:
     path_base = Path('./configs')
     if feature_type in ['resnet18', 'resnet34', 'resnet50', 'resnet101', 'resnet152']:
         path = path_base / 'resnet.yml'
+    elif feature_type in ['CLIP-ViT-B-32', 'CLIP-ViT-B-16', 'CLIP-RN50x16', 'CLIP-RN50x4',
```
and also here: `model_type: clip`.

I see that our resnets also have this issue though!

v-iashin reviewed (commented):
This was a pleasant surprise to see CLIP in PR.
I would be very excited to merge it but it requires some more effort, unfortunately.

It is all good, but here are two things that bother me the most and actually blocking the merge of these changes:

1. Lack of a down-steam task test (the `show_preds`). I think we need to implement it somehow. I see two ways: a) using Kinetics 400/Imagenet classes as Zero-shot (see how they do it for [CIFAR10](https://github.com/openai/CLIP#more-examples)) or b) with user-provided text in the config. I understand that it could be daunting but we need to provide a simple test that users could do to check that the features are working as they expect.
2. I don't like extra dependencies (`ftfy` and `clip`). I think we could merge the corresponding code of `clip` to our repo (with a proper citation/reference). Regarding `clip`, we need to decide on the `show_pred` logic first, e.g. will it be text-based (need `ftfy`) or something else?

It might be a bit too much for me to ask from you so I might do it myself but it will take some time. However, if you still have an appetite, let me know.

Again, thanks for doing this. I hope you are having a great time!

v-iashin commented on models/clip/extract_clip.py:
```diff
@@ -0,0 +1,175 @@
+from utils.utils import (action_on_extraction, form_list_from_user_input,
+                         reencode_video_with_diff_fps)
+
+try:
```
I am referring to the `try-except` chunk here.

v-iashin commented on docs/models/clip.md:
```diff
@@ -0,0 +1,103 @@
+# CLIP
+
+<figure>
```
Referring to the image tag

Kamino666 commented on configs/clip.yml:
```diff
@@ -0,0 +1,19 @@
+# Model
+feature_type: 'CLIP-ViT-B-32'
```
It's a nice idea!
But the output path and the model name may not be compatible because of slash `/` (when using `on_extraction=save_numpy`). I'm sure we can handle it. It's just a reminder XD

v-iashin replied:
Yes, a nice catch! I feel that on Windows it could potentially be a problem as well. Do you think `str.replace('/', '_')` just before saving would do?

Kamino666 replied:
Yes, it solves the problem.

Kamino666 commented on docs/models/clip.md:
````diff
@@ -0,0 +1,103 @@
+```
+
+AND install CLIP as a Python package
+```bash
````
I'm not sure whether it will work without `ftfy`. But as far as I know the CLIP code should not be difficult to embed into this repo.

Kamino666 commented:
Overall, I think the comments you made are very pertinent, but I may not have time to finish it lately. :open_mouth:

For the  reliability of the extracted feature, I have verified its effectiveness in a downstream task of the video captioning. So I may have overlooked that future users may also need to validate it. (Actually using CLIP features as visual features for video captioning task can bring a big improvement, so if you are still working on video captioning you can give it a try. I used this feature in one of my unpublished papers and cited two of your papers. If you are interested, we can talk more about this.)

I have some ideas for the future, such as setting up more ways to extract frames. Currently this repo is only able to extract frames based on FPS, while some papers extract a fixed number of frames, etc. This should not be discussed here, so I may raise an issue in the future to discuss it.

Thank you for your reply and I hope you are also having a good time.:star:

Kamino666 commented:
Have you started working on this already? I get time now.

v-iashin commented:
No, I haven't yet. Pretty much occupied during these weeks already.

Kamino666 commented:
OK, I will do this.

Kamino666 commented:
Hi, I have done these things:

1. The CLIP code is added and there is no need to install any new packages (no ftfy is OK)
2. Improved configuration of model selection (`model_name`)
3. Added a method of `show_pred`. The user provides several sentences, and gets the probabilities of each sentences. You can find the details in docs.
4. Updated the docs
5. Updated the CLIP's Colab notebook. But you may need to modify it and the ResNet one.

v-iashin commented on docs/models/clip.md:
````diff
@@ -0,0 +1,128 @@
+```bash
+python main.py \
+    feature_type=clip \
+    model_name=ViT-B/32 \
````
TODO for myself: test this on linux. 
* Will it fail because of the special characters? 
* Should we use ""?

Kamino666 replied:
There may be bugs due to some special characters, but I guess the model name in general should not have those special characters.

v-iashin replied:
It is actually ok. However, I decided to surround these arguments with `"` in the docs.

v-iashin commented on docs/models/clip.md:
````diff
@@ -0,0 +1,128 @@
+```
+Video #0: a dog smiles | a man is lifting
+frame 0: [0.987789807340838, 0.012210192659162062]
+frame 1: [0.9896199429752918, 0.010380057024708175]
````
Awesome!

v-iashin commented on docs/models/clip.md:
```diff
@@ -0,0 +1,128 @@
+    model_name=ViT-B/32 \
+    device_ids="[0, 2]" \
+    extraction_fps=1 \
+    pred_texts="['a dog smiles', 'a man is lifting']" \
```
TODO for myself: it feels like we could still benefit from showing zero-shot on kinetics classes if `pred_texts` is not defined.

v-iashin commented on main.py:
```diff
@@ -21,7 +21,7 @@ def parallel_feature_extraction(args):
     elif args.feature_type == 'vggish':
         from models.vggish.extract_vggish import ExtractVGGish  # defined here to avoid import errors
         extractor = ExtractVGGish(args)
-    elif args.feature_type in ['resnet18', 'resnet34', 'resnet50', 'resnet101', 'resnet152']:
```
Thanks for this and the rest of the resnet changes! I am happy to have them.

One comment though, please try to avoid doing multiple things in a single PR. It is considered as a bad practice since the branch is about clip and having changes of the resnet features could be misleading.

I would create a separate PR changing resnet logic instead. 

But that's ok, we can merge these two things, it is not a big deal.

Kamino666 replied:
I'm sorry :fearful:
I will avoid that next time

v-iashin replied:
No worries!

v-iashin commented on models/clip/clip_src/clip.py:
```diff
@@ -0,0 +1,233 @@
+import hashlib
```
TODO for myself: add reference to where this code was taken from.

@Kamino666 is it precisely https://github.com/openai/CLIP/blob/main/clip/clip.py without a change?

Kamino666 replied:
No, I change some codes.

v-iashin replied:
Ok, let me merge it to the `main` and we need to add this reference + a sentence that describes the changes on a high level

v-iashin commented on models/clip/clip_src/simple_tokenizer.py:
```diff
@@ -0,0 +1,132 @@
+import os
+from functools import lru_cache
+
+# import ftfy
```
But how did you make it work? 😆

Kamino666 replied:
In fact `ftfy` is just a text preprocessing tool that filters out some special characters, and since our users are unlikely to input strange characters, removing this should not be a big problem.

v-iashin commented on models/resnet/extract_resnet.py:
```diff
@@ -149,7 +150,7 @@ def _run_on_a_batch(vid_feats, batch, model, classifier, device):
         return features_with_meta
 
-    def load_model(self, device: torch.device) -> Tuple[torch.nn.Module]:
+    def load_model(self, device: torch.device) -> Tuple[torch.nn.Module, torch.nn.Module]:
```
thanks

v-iashin commented on models/resnet/extract_resnet.py:
```diff
@@ -161,18 +162,10 @@ def load_model(self, device: torch.device) -> Tuple[torch.nn.Module]:
-        else:
-            raise NotImplementedError
+        try:
+            model = getattr(models, self.model_name)
```
Neat!

v-iashin commented on utils/utils.py:
```diff
@@ -110,10 +110,14 @@ def sanity_check(args: Union[argparse.Namespace, DictConfig]):
         if args.feature_type == 'vggish':
             print('Showing class predictions is not implemented for VGGish')
+        if args.feature_type == 'clip':
+            if args.pred_texts is None:
```
TODO to myself: if I will implement zero-shot on Kinetics 400, I need to change this.

v-iashin commented on utils/utils.py:
```diff
@@ -255,3 +262,20 @@ def gpu_state_to_cpu(state_dict):
+        # some model use `/` e.g. ViT-B/16
+        real_output_path = os.path.join(real_output_path, p.replace("/", "_"))
+        real_tmp_path = os.path.join(real_tmp_path, p.replace("/", "_"))
+    args.output_path = real_output_path
```
Ok, I will need to think about this change because I am not a huge fan of patching (changing) the config after it has been loaded. If the bugs occur, it is hard to debug. However, it is just me being cautious.

Kamino666 replied:
Yes, in fact I also think there may be a more elegant way. I did that because I don't want to change other files too much.

I think it makes sense to treat the path-saving in one place. Because in the future, the newly added models may be trained on different datasets, and then new configuration items may have to be added, and the save path will need to be updated accordingly. 

e.g. model `GoodNet` has three types (`tiny` `base` `large`) and is trained on two datasets (`ImageNet`, `JFT`). One output path might be `output/GoodNet/gnet-base/imagenet`.

v-iashin replied:
True. You are right! 

Right now I am not really happy about the structure of this repo. I need to refactor it significantly (configs, models, utils) such that we could avoid so much duplication everywhere for each model which makes it hard to maintain, debug, and add new features.

v-iashin reviewed (commented):
Thanks these are great!

v-iashin commented:
Merged it to the `clip` branch. I need to test it a bit.
